src/java/org/apache/cassandra/index/sai/memory/OffheapBytes.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.memory;
+
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Offheap unsigned bytes comparable wrapper for ByteBuffer
+ */
+public class OffheapBytes implements Comparable<OffheapBytes>
+{
+    public final ByteBuffer buffer;
+
+    public OffheapBytes(byte[] bytes)
+    {
+        this(new BytesRef(bytes));
+    }
+
+    public OffheapBytes(BytesRef bytesRef)
+    {
+        this.buffer = ByteBuffer.allocateDirect(bytesRef.length);
```

Comment:
I'm not sure if I understand the purpose of using direct byte buffers here the way they are used. 

It seems we're not really pooling / reusing those buffers anywhere, and just allocating short-lived buffers on-demand, e.g. in `BlockIndex#rangeMatch`.

Problems:
* Allocating direct buffers is significantly slower than allocating on-heap buffers.
* It is possible that eventually you'll run out of off-heap memory and that would trigger a full GC; possibly blocking the other important stuff for much longer than minor per-segment heap collections done incrementally by G1 / ZGC.

Reply:
See `MemtableTrie`